### PR TITLE
[MRG] adding softmax layer

### PR DESCRIPTION
**Pull request by mrityunjay-tripathi**
This should close PR #1958 worked by @sreenikSS.

### CODE REVIEW

**Comment by ojhalakshya** [comment]
Hi @mrityunjay-tripathi 
could you also add tests for this layer.
Thanks.

**Comment by mrityunjay-tripathi** [comment]
>Hi @mrityunjay-tripathi
could you also add tests for this layer.
Thanks.

I am working on this @ojhalakshya. I will commit them as soon as it completes :)

**Comment by jeffin143** [comment]
@mrityunjay-tripathi : Thanks you for the PR ❤️ , But I guess we should Pull @sreenikSS branch and the continue work upon it, Because it has only left few comments to be resolved, So I guess we should add his commit too :)

**Comment by mrityunjay-tripathi** [comment]
Hi @jeffin143, There was a branching issue in that branch. I was a little hesitant about what to do, so I opened new PR. Also, the current implementation seems incorrect. I tested it locally and I got the following results:
Code:
```
#include <iostream>
#include <mlpack/core.hpp>
#include <mlpack/methods/ann/layer/softmax.hpp>

using namespace std;
using namespace mlpack::ann;

int main()
{
  arma::mat input = 20 * arma::randu<arma::mat>(1, 10);
  arma::mat output;
  
  SoftMax<> module;
  module.Forward(input, output);
  
  input.print("Input: ");
  output.print("Output: ");
}
```
Output:
```
Input: 
   15.7364    5.0096   14.2134   18.9334    0.3854    8.0980    5.0264    0.4542   10.4129    6.8934
Output: 
   0.9772   0.9316   0.9748   0.9809   0.5116   0.9565   0.9318   0.5525   0.9659   0.9493
```

Clearly, this is not something we want. The approximation of e^x seems to have the problem.

**Comment by sreenikSS** [comment]
@jeffin143 it's alright. @mrityunjay-tripathi there was a small change required in the Backward() function also if I remember correctly.

**Comment by jeffin143** [comment]
If it's find by you both , I am more than happy :)

**Comment by mrityunjay-tripathi** [comment]
>If it's find by you both , I am more than happy :)

I am really helpless, it is showing "unknown repository". Probably @sreenikSS has deleted that branch.

**Comment by mrityunjay-tripathi** [comment]
>@jeffin143 it's alright. @mrityunjay-tripathi there was a small change required in the Backward() function also if I remember correctly.

I have to go through that.

**Comment by mrityunjay-tripathi** [comment]
Hi @sreenikSS, can you provide the reference for the approximation of e^-x. As of now, the forward function returns the correct result but it will be better if it is efficient as well.

**Comment by sreenikSS** [comment]
@mrityunjay-tripathi you can check the ```log_softmax_impl.hpp``` for the approximation of exp.

**Comment by mrityunjay-tripathi** [comment]
Hey @birm @rcurtin @zoq @favre49, it would be a great help if someone can explain to me how the backward function of a layer has to be implemented whose output is dependent on all of the input (like softmax function, i.e. e^x_i / sum(e^x_i)). Since the derivative of a vector wrt. another vector returns a Jacobian Matrix, but if the input is a vector then forward and backward both have to return a vector!?

With some pen-paper work, I managed to write an expression which passes on the Jacobian Test but I doubt if it is correct. Also, the values match with that of PyTorch's softmax function.

In PyTorch:
```
import torch
import torch.nn as nn

input = torch.tensor([1.7, 3.6, 4.9], requires_grad = True)
output = nn.Softmax(dim = 0)(input)

print("Forward:", output)
t = torch.zeros(3)
t[0] = 1
output.backward(t)

print("Backward:", input.grad)
```
Output:
```
Forward: tensor([0.0310, 0.2075, 0.7614], grad_fn=<SoftmaxBackward>)
Backward: tensor([ 0.0301, -0.0064, -0.0236])
```

In this PR:
```
#include <iostream>
#include <mlpack/core.hpp>
#include <mlpack/methods/ann/layer/softmax.hpp>

using namespace std;
using namespace mlpack::ann;
int main()
{
  arma::mat input = arma::mat("1.7; 3.6; 4.9");
  arma::mat output, gy = arma::zeros(3, 1), g;
  gy(0) = 1;
  SoftMax<> module;
  module.Forward(input, output);
  output.print("Forward:");
  module.Backward(output, gy, g);
  g.print("Backward:");
}
```
Output:
```
Forward:
   0.0310
   0.2075
   0.7614
Backward:
   0.0301
  -0.0064
  -0.0236

```
It would be extremely helpful if you review it once.

File: src/mlpack/methods/ann/layer/softmax.hpp [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+
+ private:
+  //! Locally-stored delta object.
+  OutputDataType delta;
+
+  //! Locally-stored output parameter object.
+  OutputDataType outputParameter;
+
+  //! Locally-stored value of the axis along which softmax is calculated.
+  bool axis;
```
Comment (mrityunjay-tripathi) [inline src/mlpack/methods/ann/layer/softmax.hpp:96]: ohh... I need to remove this (it's not used).

File: src/mlpack/methods/ann/layer/softmax_impl.hpp [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+{
+  // Nothing to do here.
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename InputType, typename OutputType>
+void SoftMax<InputDataType, OutputDataType>::Forward(
+    const InputType& input, OutputType& output)
+{
+  output.set_size(arma::size(input));
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/softmax_impl.hpp:33]: Since you assign the output in line 37, no need to set the size here.

File: src/mlpack/methods/ann/layer/softmax_impl.hpp [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+template<typename eT>
+void SoftMax<InputDataType, OutputDataType>::Backward(
+    const arma::Mat<eT>& input,
+    const arma::Mat<eT>& gy,
+    arma::Mat<eT>& g)
+{
+  g.set_size(input.n_rows, input.n_cols);
+  for (size_t i = 0; i < input.n_cols; ++i)
+  {
+    const arma::Mat<eT> gyi = gy.col(i);
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/softmax_impl.hpp:52]: Not sure that is necessary, you could directly use `gy.col(i)` and `input.col(i)`.

File: src/mlpack/methods/ann/layer/softmax_impl.hpp [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+    const arma::Mat<eT> xi = input.col(i);
+
+    g.col(i) = xi % (gyi - arma::repmat(gyi.t() * xi, gyi.n_rows, gyi.n_cols));
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/softmax_impl.hpp:55]: Looks like `each_col` could work as well, and would avoid the extra matrix.

File: src/mlpack/methods/ann/layer/softmax.hpp [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+
+#include <mlpack/prereqs.hpp>
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+/**
+ * Implementation of the softmax layer. The softmax loss layer computes
+ * the multinomial logistic loss of the softmax of its inputs. It should be
+ * used for inference only and not with NLL loss (use logsoftmax instead).
```
[the remark this answers is not in the document]
Comment (mrityunjay-tripathi) [reply src/mlpack/methods/ann/layer/softmax.hpp]: Done.

File: src/mlpack/methods/ann/layer/softmax.hpp [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+ * @tparam InputDataType Type of the input data (arma::colvec, arma::mat,
+ *         arma::sp_mat or arma::cube).
+ * @tparam OutputDataType Type of the output data (arma::colvec, arma::mat,
+ *         arma::sp_mat or arma::cube).
+ */
+template <
+    typename InputDataType = arma::mat,
+    typename OutputDataType = arma::mat
+>
+class SoftMax
```
[the remark this answers is not in the document]
Comment (mrityunjay-tripathi) [reply src/mlpack/methods/ann/layer/softmax.hpp]: Agree. Also `Softmax` looks nicer :)

File: src/mlpack/methods/ann/layer/softmax_impl.hpp (hunk 3 above)
Comment (mrityunjay-tripathi) [reply src/mlpack/methods/ann/layer/softmax_impl.hpp]: Done.

File: src/mlpack/methods/ann/layer/softmax_impl.hpp (hunk 2 above)
Comment (mrityunjay-tripathi) [reply src/mlpack/methods/ann/layer/softmax_impl.hpp]: Done but I wonder how armadillo does this? Is it not that when we initialize the matrix with predefined shape (static allocation), it accesses the elements faster?

File: src/mlpack/methods/ann/layer/softmax_impl.hpp (hunk 4 above)
Comment (mrityunjay-tripathi) [reply src/mlpack/methods/ann/layer/softmax_impl.hpp]: From the documentation, I couldn't clearly understand how to achieve this as the right-hand side of the expression depends on the ith column of `input` and `gy`.

Comment (mrityunjay-tripathi) [reply src/mlpack/methods/ann/layer/softmax_impl.hpp]: One more question, should I use `arma::as_scalar` or `arma::dot` [here](https://github.com/mlpack/mlpack/pull/2351/commits/b7f318ef608a42a204c8b96553fe9da8785a36a5), as on some tests `as_scalar` is fast and on other `dot` is fast?

File: src/mlpack/methods/ann/layer/softmax_impl.hpp [hunk 7]
```diff
@@ -44,12 +43,7 @@ void Softmax<InputDataType, OutputDataType>::Backward(
...
     arma::Mat<eT>& g)
 {
-  g.set_size(input.n_rows, input.n_cols);
-  for (size_t i = 0; i < input.n_cols; ++i)
-  {
-    g.col(i) = input.col(i) % (gy.col(i)
-        - arma::as_scalar(gy.col(i).t() * input.col(i)));
-  }
+  g = input % (gy - arma::repmat(arma::sum(gy % input), input.n_rows, 1));
 }
```
Comment (mrityunjay-tripathi) [inline src/mlpack/methods/ann/layer/softmax_impl.hpp:47]: @zoq: I think this is fine now. What do you think?

File: .gitignore [hunk 8]
```diff
@@ -1,6 +1,7 @@
 build*
 xcode*
 .DS_Store
+.vscode
 src/mlpack/core/util/gitversion.hpp
 src/mlpack/core/util/arma_config.hpp
```
Comment (mrityunjay-tripathi) [inline .gitignore:6]: I've added `.vscode` since it becomes annoying sometimes if vscode files get added by mistake. I think its fine to add this. Or should I open a separate PR for this?

Comment (zoq) [reply .gitignore:6]: That's fine.

File: src/mlpack/methods/ann/layer/softmax_impl.hpp (hunk 7 above)
Comment (zoq) [reply src/mlpack/methods/ann/layer/softmax_impl.hpp:47]: Looks good.

File: src/mlpack/methods/ann/layer/layer_types.hpp [hunk 9]
```diff
@@ -233,6 +234,7 @@ using LayerTypes = boost::variant<
     Linear<arma::mat, arma::mat, NoRegularizer>*,
     LinearNoBias<arma::mat, arma::mat, NoRegularizer>*,
     LogSoftMax<arma::mat, arma::mat>*,
+    Softmax<arma::mat, arma::mat>*,
```
Comment (lozhnikov) [inline src/mlpack/methods/ann/layer/layer_types.hpp:237]: Other layers are in alphabetical order. Could you move the softmax layer accordingly?

Comment (mrityunjay-tripathi) [reply src/mlpack/methods/ann/layer/layer_types.hpp]: Thanks, @lozhnikov for the review. I've made the required changes. I noticed some other LayerTypes were also out of order, so arranged them as well.

File: src/mlpack/methods/ann/layer/softmax.hpp [hunk 10]
```diff
@@ -0,0 +1,100 @@
+/**
+ * @file softmax.hpp
```
Comment (birm) [inline src/mlpack/methods/ann/layer/softmax.hpp:2]:
```suggestion
 * @file methods/ann/layer/softmax.hpp
```

File: src/mlpack/methods/ann/layer/softmax_impl.hpp [hunk 11]
```diff
@@ -0,0 +1,61 @@
+/**
+ * @file softmax_impl.hpp
```
Comment (birm) [inline src/mlpack/methods/ann/layer/softmax_impl.hpp:2]:
```suggestion
 * @file methods/ann/layer/softmax_impl.hpp
```

**Comment by mrityunjay-tripathi** [comment]
Thanks, @birm for the review.
>Good to have a long-running PR chain resolved :)

:smiley:

**Comment by ShikharJ** [comment]
Please wait on the merge, I'd like to review this as well.

File: src/mlpack/methods/ann/layer/layer_types.hpp [hunk 12]
```diff
@@ -259,8 +258,11 @@ using LayerTypes = boost::variant<
     NegativeLogLikelihood<arma::mat, arma::mat>*,
     Padding<arma::mat, arma::mat>*,
     PReLU<arma::mat, arma::mat>*,
+    Softmax<arma::mat, arma::mat>*,
+    TransposedConvolution<NaiveConvolution<ValidConvolution>,
+            NaiveConvolution<ValidConvolution>,
+            NaiveConvolution<ValidConvolution>, arma::mat, arma::mat>*,
     WeightNorm<arma::mat, arma::mat>*,
-    CELU<arma::mat, arma::mat>*,
     MoreTypes,
```
Comment (ShikharJ) [inline src/mlpack/methods/ann/layer/layer_types.hpp:266]: Can you please restore the import re-arrangements made in this file? These are redundant.

Comment (mrityunjay-tripathi) [reply src/mlpack/methods/ann/layer/layer_types.hpp:266]: @ShikharJ: I rearranged the `LayerTypes` to maintain the alphabetical order as suggested by Mikhail. https://github.com/mlpack/mlpack/pull/2351#discussion_r423131788

File: src/mlpack/methods/ann/layer/softmax.hpp [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+   *
+   * @param input Input data used for evaluating the specified function.
+   * @param output Resulting output activation.
+   */
+  template<typename InputType, typename OutputType>
+  void Forward(const InputType& input, OutputType& output);
+
+  /**
+   * Ordinary feed backward pass of a neural network, calculating the function
+   * f(x) by propagating x backwards trough f. Using the results from the feed
```
Comment (ShikharJ) [inline src/mlpack/methods/ann/layer/softmax.hpp:58]: Typo here, it should be `through`.

Comment (mrityunjay-tripathi) [reply src/mlpack/methods/ann/layer/softmax.hpp]: Corrected.

**Comment by ShikharJ** [comment]
Hmm I'm wondering whether this should be exposed as a layer or rather as an activation function.

**Comment by kartikdutt18** [comment]
> Hmm I'm wondering whether this should be exposed as a layer or rather as an activation function.

Hey @ShikharJ, Since this function doesn't require any parameters maybe it makes sense to implement it in the activation functions folder since the only those activation functions are implemented as layers in which the user would need to have access to some parameters. Since that is not the case with this one maybe it makes sense to have softmax activation function added in base layer. Let me know what you think.

**Comment by mrityunjay-tripathi** [comment]
Hey @ShikharJ @kartikdutt18, I think the main problem is the backward function, where general implementation in the base layer is like `g = gy % derivative` but it's not like that in softmax. I can produce more tests if you say to check if this implementation is correct. And the other thing is that softmax is a "normalized" exponential function so `softmax(x_i)` depends on each of the `x_j` (j=1, 2,3,...,n) which is not the case with other activation functions. The `log_softmax` is also implemented as a layer (I think for that reason only). Maybe @zoq can help.

**Comment by zoq** [comment]
When I think about Softmax I rather see it as a layer than an activation function, so for me it would make sense to implement it as a layer and put it into the layer folder.

**Comment by kartikdutt18** [comment]
> When I think about Softmax I rather see it as a layer than an activation function, so for me it would make sense to implement it as a layer and put it into the layer folder.

Great, I think this is ready then. Thanks a lot for the clarification.
